File: workspace/vat.py
"""Internal VAT review worksheets. No filing adapter or automatic tax eligibility."""
import calendar
import hashlib
import json
import uuid
from collections import defaultdict
from datetime import date
from decimal import Decimal, ROUND_HALF_UP

from django.db import transaction
from django.utils import timezone
from server import integer, label, money, require
from .models import Company, Payment, Purchase, Sale, SupplierPayment, VatWorksheet
from .purchases import calendar_date
from .services import scoped


def totals(sheet):
    reviewed = [l for l in sheet.lines if l['reviewed']]
    sales = sum(l['retained_tax'] for l in reviewed if l['direction'] == 'sale')
    purchases = sum(l['retained_tax'] for l in reviewed if l['direction'] == 'purchase')
    net = sales - purchases - sheet.opening_credit
    return dict(sales_tax=sales, deductible_tax=purchases, opening_credit=sheet.opening_credit,
                payable=max(net, 0), carryforward=max(-net, 0), pending=sum(not l['reviewed'] for l in sheet.lines))


def fingerprint(data):
    return hashlib.sha256(json.dumps(data, sort_keys=True, ensure_ascii=True, separators=(',', ':')).encode()).hexdigest()


class VatService:
    def __init__(self, company, actor):
        self.company, self.actor = company, actor

    def handle(self, action, p):
        return {'vat-create': self.create, 'vat-import': self.import_sources, 'vat-line': self.line,
                'vat-remove': self.remove, 'vat-settings': self.settings, 'vat-approve': self.approve,
                'vat-void': self.void}[action](p)

    def draft(self, p):
        sheet = scoped(VatWorksheet, self.company, p.get('id'))
        require(sheet.status == 'draft', 'Cette préparation est verrouillée. Annulez son approbation pour préparer une nouvelle version.')
        require(integer(p.get('version'), 'Version', 1) == sheet.version, 'La préparation a changé. Actualisez la page.')
        return sheet

    def save(self, sheet):
        sheet.version += 1
        sheet.save()
        return {'id': sheet.pk, 'version': sheet.version}

    def create(self, p):
        start = calendar_date(p.get('start'), 'Début de période')
        cadence = p.get('cadence')
        require(cadence in ('monthly', 'quarterly'), 'Périodicité invalide.')
        require(start.day == 1 and (cadence == 'monthly' or start.month in (1, 4, 7, 10)), 'Choisissez le premier jour du mois ou du trimestre.')
        month = start.month if cadence == 'monthly' else start.month + 2
        end = date(start.year, month, calendar.monthrange(start.year, month)[1])
        require(not VatWorksheet.objects.filter(company=self.company, start__lte=end, end__gte=start).exclude(status__in=['voided', 'discarded']).exists(),
                'Une préparation active couvre déjà cette période.')
        sales_basis, purchase_basis = p.get('sales_basis'), p.get('purchase_basis')
        require(sales_basis in ('invoices', 'payments') and purchase_basis in ('invoices', 'payments'), 'Choisissez les bases d’import des ventes et achats.')
        sheet = VatWorksheet.objects.create(company=self.company, start=start, end=end, cadence=cadence,
            sales_basis=sales_basis, purchase_basis=purchase_basis, created_by=self.actor)
        return {'id': sheet.pk, 'version': sheet.version}

    def sources(self, sheet):
        """Candidate amounts, never a determination of deductibility/tax liability."""
        results = {}
        for direction, model, payment_model, fk, status, basis in [
            ('sale', Sale, Payment, 'sale_id', 'active', sheet.sales_basis),
            ('purchase', Purchase, SupplierPayment, 'purchase_id', 'received', sheet.purchase_basis)]:
            payments = defaultdict(list)
            for payment in payment_model.objects.filter(company=self.company).order_by('created_at' if direction == 'sale' else 'payment_date', 'id'):
                payments[getattr(payment, fk)].append(payment)
            for invoice in model.objects.filter(company=self.company, status=status).select_related('customer' if direction == 'sale' else 'supplier'):
                if not invoice.total:
                    continue
                day = timezone.localdate(invoice.created_at) if direction == 'sale' else invoice.invoice_date
                party = invoice.customer if direction == 'sale' else invoice.supplier
                common = dict(document_key=f'{direction}:{invoice.pk}', document_tax=invoice.tax,
                    invoice_date=day.isoformat(), invoice_total=invoice.total,
                    reference=f'VT-{invoice.number:04d}' if direction == 'sale' else invoice.supplier_reference,
                    party=invoice.customer_name if direction == 'sale' else invoice.supplier_name,
                    fiscal_id=getattr(party, 'fiscal_id', ''), ice=getattr(party, 'ice', ''), direction=direction)
                if basis == 'invoices':
                    events = [(f'{direction}:{invoice.pk}', day, invoice.subtotal, invoice.tax, '', '')]
                else:
                    events, cumulative = [], 0
                    def proportional(value, paid):
                        return int((Decimal(value) * paid / invoice.total).quantize(Decimal('1'), rounding=ROUND_HALF_UP))
                    for payment in payments[invoice.pk]:
                        payment_day = timezone.localdate(payment.created_at) if direction == 'sale' else payment.payment_date
                        before = cumulative
                        cumulative += payment.amount
                        require(cumulative <= invoice.total, 'Les règlements d’une facture dépassent son total. Corrigez la source avant l’import TVA.')
                        tax = proportional(invoice.tax, cumulative) - proportional(invoice.tax, before)
                        # TTC - allocated tax preserves the exact paid amount in centimes.
                        net = payment.amount - tax
                        events.append((f'{direction}-payment:{payment.pk}', payment_day, net, tax, payment.method, getattr(payment, 'reference', '')))
                for key, source_day, net, tax, method, payment_reference in events:
                    if sheet.start <= source_day <= sheet.end:
                        snapshot = dict(common, key=key, date=source_day.isoformat(), net=net, tax=tax,
                                        method=method, payment_reference=payment_reference, basis=basis)
                        results[key] = snapshot
        return results

    def import_sources(self, p):
        sheet = self.draft(p)
        sources = self.sources(sheet)
        existing = {l['key']: l for l in sheet.lines if l['source']}
        added = changed = 0
        for key, snapshot in sources.items():
            prior = existing.get(key)
            if prior and prior['source'] == snapshot:
                continue
            row = dict(key=key, direction=snapshot['direction'], date=snapshot['date'], reference=snapshot['reference'],
                party=snapshot['party'], net=snapshot['net'], tax=snapshot['tax'], retained_tax=0,
                reviewed=False, reason='', source=snapshot)
            if prior:
                sheet.lines[sheet.lines.index(prior)] = row
                changed += 1
            else:
                sheet.lines.append(row)
                added += 1
        require(len(sheet.lines) <= 2000, 'Maximum : 2 000 lignes par préparation.')
        result = self.save(sheet)
        return dict(result, added=added, changed=changed)

    def line(self, p):
        sheet = self.draft(p)
        key = p.get('key')
        prior = next((l for l in sheet.lines if l['key'] == key), None) if key else None
        require(not key or prior is not None, 'Ligne introuvable dans cette préparation.')
        if prior and prior['source']:
            require(self.sources(sheet).get(key) == prior['source'], 'La source a changé. Réimportez les sources ou retirez la ligne obsolète.')
            row = dict(prior)
        else:
            direction = p.get('direction')
            require(direction in ('sale', 'purchase'), 'Choisissez vente ou achat.')
            day = calendar_date(p.get('date'), 'Date de prise en compte')
            require(sheet.start <= day <= sheet.end, 'La date doit appartenir à la période.')
            row = dict(key=key or f'manual:{uuid.uuid4()}', direction=direction, date=day.isoformat(),
                reference=label(p.get('reference'), 'Référence', limit=80), party=label(p.get('party'), 'Tiers'),
                net=money(p.get('net'), 'Montant HT'), tax=money(p.get('tax'), 'TVA du document'), source=None)
        retained = money(p.get('retained_tax'), 'TVA retenue')
        require(retained <= row['tax'], 'La TVA retenue ne peut pas dépasser la TVA source.')
        reason = label(p.get('reason'), 'Justification du traitement', limit=300)
        row.update(retained_tax=retained, reviewed=True, reason=reason)
        if prior:
            sheet.lines[sheet.lines.index(prior)] = row
        else:
            require(len(sheet.lines) < 2000, 'Maximum : 2 000 lignes par préparation.')
            sheet.lines.append(row)
        return dict(self.save(sheet), key=row['key'])

    def remove(self, p):
        sheet = self.draft(p)
        key = p.get('key')
        require(any(l['key'] == key for l in sheet.lines), 'Ligne introuvable.')
        reason = label(p.get('reason'), 'Motif', limit=300)
        sheet.lines = [l for l in sheet.lines if l['key'] != key]
        return dict(self.save(sheet), key=key, reason=reason)

    def settings(self, p):
        sheet = self.draft(p)
        sheet.opening_credit = money(p.get('opening_credit', 0), 'Crédit antérieur')
        sheet.credit_note = label(p.get('credit_note', ''), 'Origine du crédit', bool(sheet.opening_credit), 300)
        sheet.note = label(p.get('note', ''), 'Note de préparation', False, 1000)
        return self.save(sheet)

    def diagnostics(self, sheet):
        sources = self.sources(sheet)
        stale = [l['key'] for l in sheet.lines if l['source'] and sources.get(l['key']) != l['source']]
        known = {l['key'] for l in sheet.lines if l['source']}
        missing = [k for k in sources if k not in known]
        return dict(stale=stale, missing=missing)

    def approve(self, p):
        sheet = self.draft(p)
        require(totals(sheet)['pending'] == 0, 'Vérifiez chaque ligne avant l’approbation.')
        require(p.get('confirmed') is True, 'Confirmez la vérification du périmètre, des montants et du crédit antérieur.')
        checks = self.diagnostics(sheet)
        require(not checks['stale'], 'Des sources ont changé ou disparu. Réimportez ou retirez les lignes obsolètes.')
        require(not checks['missing'], 'De nouvelles sources sont disponibles. Importez-les et vérifiez leur traitement, même si leur TVA retenue est nulle.')
        used_keys, allocated = set(), defaultdict(int)
        for other in VatWorksheet.objects.filter(company=self.company, status='approved'):
            for line in other.lines:
                if line['source'] and line['retained_tax']:
                    used_keys.add(line['key'])
                    allocated[line['source']['document_key']] += line['retained_tax']
        for line in sheet.lines:
            if not line['source'] or not line['retained_tax']:
                continue
            require(line['key'] not in used_keys, 'Une source est déjà retenue dans une autre préparation approuvée.')
            key = line['source']['document_key']
            allocated[key] += line['retained_tax']
            require(allocated[key] <= line['source']['document_tax'], 'Le cumul de TVA retenue dépasse la TVA de la facture dans les préparations approuvées.')
        sheet.company_snapshot = {k: getattr(self.company, k) for k in ('name', 'fiscal_id', 'ice', 'address')}
        sheet.approved_by, sheet.approved_at = self.actor, timezone.now()
        sheet.status = 'approved'
        sheet.approved_snapshot = dict(format='atlas-vat-review', version=1, worksheet_id=sheet.pk,
            company_id=self.company.pk, company=sheet.company_snapshot, start=sheet.start.isoformat(), end=sheet.end.isoformat(),
            cadence=sheet.cadence, sales_basis=sheet.sales_basis, purchase_basis=sheet.purchase_basis,
            lines=sheet.lines, totals=totals(sheet), credit_note=sheet.credit_note, note=sheet.note,
            approved_by=self.actor.pk, approved_at=sheet.approved_at.isoformat(), filing_status='not_submitted')
        sheet.checksum = fingerprint(sheet.approved_snapshot)
        return self.save(sheet)

    def void(self, p):
        sheet = scoped(VatWorksheet, self.company, p.get('id'))
        require(sheet.status in ('draft', 'approved'), 'Cette préparation est déjà annulée.')
        require(integer(p.get('version'), 'Version', 1) == sheet.version, 'La préparation a changé. Actualisez la page.')
        sheet.void_reason = label(p.get('reason'), 'Motif', limit=300)
        sheet.status, sheet.voided_by, sheet.voided_at = ('voided' if sheet.status == 'approved' else 'discarded'), self.actor, timezone.now()
        return self.save(sheet)

    @transaction.atomic
    def state(self):
        Company.objects.select_for_update().get(pk=self.company.pk)
        rows = []
        for sheet in VatWorksheet.objects.filter(company=self.company).order_by('-start', '-id'):
            row = {f.attname: getattr(sheet, f.attname) for f in sheet._meta.fields}
            row['totals'] = totals(sheet)
            row['diagnostics'] = self.diagnostics(sheet) if sheet.status == 'draft' else None
            rows.append(row)
        return {'worksheets': rows}

File: web/vat.js
'use strict';
window.AtlasVat = {
  selected:null,saving:false,key:null,
  reset(){this.selected=null;},
  data(){return state.vat.worksheets;},
  current(){return this.data().find(s=>s.id===this.selected);},
  can(){return AtlasPlatform.can('vat_write');},
  name(s){return s.cadence==='quarterly'?`T${Math.floor((Number(s.start.slice(5,7))-1)/3)+1} · ${s.start.slice(0,4)}`:`${date(s.start+'T12:00:00',{month:'long',year:'numeric'})}`;},
  status(s){return badge({draft:'À préparer',approved:'Approuvée en interne',voided:'Approbation annulée',discarded:'Abandonnée'}[s.status],s.status==='approved'?'':s.status==='draft'?'warning':'neutral');},
  button(action,text,id='',primary=false){return `<button type="button" class="btn ${primary?'primary':''}" data-action="vat-${action}" data-id="${id}">${esc(text)}</button>`;},
  basis(value){return value==='invoices'?'Factures':'Règlements';},
  view(){
    const s=this.current();
    return heading('PRÉPARER, VÉRIFIER, CONSERVER','Votre TVA','Une préparation par société, avec des montants justifiés et une revue explicite.',this.can()?this.button('new','Nouvelle préparation','',true):'')+
      `<div class="notice">Préparation interne pour revue comptable. Le choix des sources ne détermine pas le régime fiscal ni la déductibilité. L’export XML DGI et le dépôt ne sont pas encore disponibles.</div>`+(s?this.detail(s):this.list());
  },
  list(){
    const rows=this.data();
    return `<section class="card vat-list">${rows.length?`<div class="table-scroll"><table><thead><tr><th>Période</th><th>État</th><th>Lignes à vérifier</th><th class="amount">TVA à payer estimée</th><th></th></tr></thead><tbody>${rows.map(s=>`<tr><td><strong>${esc(this.name(s))}</strong><small class="purchase-sub">${s.start} → ${s.end}</small></td><td>${this.status(s)}</td><td>${s.totals.pending}</td><td class="amount">${currency(s.totals.payable)}</td><td>${this.button('open','Ouvrir',s.id)}</td></tr>`).join('')}</tbody></table></div>`:empty('Aucune préparation TVA','Créez une période, importez les sources puis vérifiez chaque montant.','','',true)}</section>`;
  },
  detail(s){
    const draft=s.status==='draft',edit=draft&&this.can(),t=s.totals,checks=s.diagnostics;
    return `<div class="toolbar vat-toolbar">${this.button('back','Toutes les périodes')}<h2>${esc(this.name(s))}</h2>${this.status(s)}</div><section class="stats-grid vat-stats">${statsCard('TVA collectée retenue',amount(t.sales_tax),'DH','Lignes vérifiées uniquement','sale',true)}${statsCard('TVA déductible retenue',amount(t.deductible_tax),'DH','Traitement validé ligne par ligne','check')}${statsCard('À payer estimé',amount(t.payable),'DH',`${t.pending} lignes restant à vérifier`,'wallet')}${statsCard('Crédit restant estimé',amount(t.carryforward),'DH','Report à confirmer dans la période suivante','arrows')}</section><section class="card vat-context"><div><strong>Sources proposées</strong><p>Ventes : ${this.basis(s.sales_basis)} · Achats : ${this.basis(s.purchase_basis)}<br>${s.start} → ${s.end}</p></div><div><strong>Crédit antérieur : ${currency(s.opening_credit)}</strong><p>${esc(s.credit_note||'Aucun crédit reporté')}</p></div>${edit?this.button('settings','Crédit & notes',s.id):''}</section>${t.pending?`<div class="notice">Calcul partiel : ${t.pending} lignes ne contribuent pas encore aux totaux. Une TVA retenue à zéro doit aussi être justifiée.</div>`:''}${checks?.stale.length?`<div class="notice">${checks.stale.length} sources ont changé ou disparu. Réimportez les sources modifiées et retirez les lignes devenues obsolètes.</div>`:''}${checks?.missing.length?`<div class="notice">${checks.missing.length} sources disponibles à importer avant approbation.</div>`:''}<div class="toolbar vat-toolbar"><p class="stat-note">${s.lines.length} lignes · montants en MAD</p><div class="vat-actions">${edit?`${this.button('import','Importer / actualiser',s.id)}${this.button('manual','Ajouter une ligne',s.id)}${this.button('approve','Approuver la préparation',s.id,true)}`:''}${s.status==='approved'&&AtlasPlatform.can('export')?`<a class="btn" href="/api/companies/${AtlasPlatform.companyId}/vat/export?id=${s.id}&format=csv">Exporter CSV</a><a class="btn" href="/api/companies/${AtlasPlatform.companyId}/vat/export?id=${s.id}&format=json">Télécharger l’instantané</a>`:''}</div></div><section class="card">${s.lines.length?`<div class="table-scroll"><table><thead><tr><th>Document / tiers</th><th>Origine</th><th class="amount">HT proposé</th><th class="amount">TVA source</th><th class="amount">TVA retenue</th><th>Revue</th><th></th></tr></thead><tbody>${s.lines.map(l=>`<tr data-vat-key="${esc(l.key)}"><td><strong>${esc(l.reference)}</strong><small class="purchase-sub">${esc(l.party)} · ${l.date}</small></td><td>${l.direction==='sale'?'Vente':'Achat'}<small class="purchase-sub">${l.source?this.basis(l.source.basis):'Manuelle'}</small></td><td class="amount">${currency(l.net)}</td><td class="amount">${currency(l.tax)}</td><td class="amount">${l.reviewed?currency(l.retained_tax):'—'}</td><td>${badge(l.reviewed?'Vérifiée':'À vérifier',l.reviewed?'':'warning')}${checks?.stale.includes(l.key)?'<small class="purchase-sub">Source obsolète</small>':''}</td><td><button class="btn" data-action="vat-line" data-key="${esc(l.key)}">${edit?'Vérifier':'Consulter'}</button></td></tr>`).join('')}</tbody></table></div>`:empty('Aucune ligne','Importez les sources de la période ou saisissez une ligne manuelle.','','',true)}</section>${s.note?`<div class="notice">${esc(s.note)}</div>`:''}${s.checksum?`<div class="vat-approval"><strong>Instantané approuvé le ${date(s.approved_at)}</strong><p>Société : ${esc(s.company_snapshot.name)} · IF : ${esc(s.company_snapshot.fiscal_id||'Non renseigné')} · ICE : ${esc(s.company_snapshot.ice||'Non renseigné')}</p><small>SHA-256 : ${esc(s.checksum)}</small><p>Approbation interne · aucune déclaration déposée par Atlas.</p></div>`:''}${s.void_reason?`<div class="notice">Motif d’annulation : ${esc(s.void_reason)}</div>`:''}${this.can()&&['draft','approved'].includes(s.status)?`<div class="vat-footer">${this.button('void',draft?'Abandonner la préparation':'Annuler l’approbation',s.id)}</div>`:''}`;
  },
  hidden(s){return `<input type="hidden" name="id" value="${s.id}"><input type="hidden" name="version" value="${s.version}">`;},
  form(title,subtitle,body,action,submit,wide=false){modal(title,subtitle,body,`${cancelButton}<button type="submit" class="btn primary">${submit}</button>`,wide,`vat-${action}-form`);this.key=crypto.randomUUID();},
  create(){
    const basis=(label,name)=>AtlasAccounting.select(label,name,[{id:'payments',name:'Règlements de la période'},{id:'invoices',name:'Factures de la période'}],'payments');
    this.form('Une préparation TVA','Sélectionnez la période et les sources à examiner.',`<div class="form-grid">${field('Premier jour de la période','start',AtlasPurchases.today().slice(0,7)+'-01','date')}${AtlasAccounting.select('Périodicité','cadence',[{id:'monthly',name:'Mensuelle'},{id:'quarterly',name:'Trimestrielle'}],'monthly')}${basis('Base d’import des ventes','sales_basis')}${basis('Base d’import des achats','purchase_basis')}</div><div class="notice">Ces bases servent à proposer les sources. Pour un règlement partiel, la TVA est répartie proportionnellement au TTC, avec arrondi cumulé au centime. Vérifiez le traitement applicable, notamment en présence de plusieurs taux ou de retenues.</div>`,'create','Créer la préparation',true);
  },
  line(key){
    const s=this.current(),l=s.lines.find(l=>l.key===key),edit=s.status==='draft'&&this.can();
    if(!edit){modal('Détail TVA',l.reference,`<div class="notice">${esc(l.party)} · ${l.date}<br>HT ${currency(l.net)} · TVA source ${currency(l.tax)} · TVA retenue ${currency(l.retained_tax)}<br>${esc(l.reason)}</div>`,cancelButton);return;}
    const source=l?.source;
    const fields=source?`<div class="notice">${esc(l.reference)} · ${esc(l.party)}<br>Date facture : ${source.invoice_date} · Date source : ${source.date}<br>HT proposé ${currency(l.net)} · TVA source ${currency(l.tax)}<br>${esc(source.method)} ${esc(source.payment_reference)}<br>IF : ${esc(source.fiscal_id||'Non renseigné')} · ICE : ${esc(source.ice||'Non renseigné')}</div>`:`<div class="form-grid">${AtlasAccounting.select('Nature','direction',[{id:'sale',name:'Vente / TVA collectée'},{id:'purchase',name:'Achat / TVA déductible'}],l?.direction||'purchase')}${field('Date de prise en compte','date',l?.date||s.start,'date')}${field('Référence du document','reference',l?.reference||'')}${field('Tiers','party',l?.party||'')}${field('Montant HT','net',((l?.net||0)/100).toFixed(2),'number',true,false,'min="0" max="10000000" step="0.01"')}${field('TVA du document','tax',((l?.tax||0)/100).toFixed(2),'number',true,false,'min="0" max="10000000" step="0.01"')}</div>`;
    this.form(l?'Vérifier la ligne':'Une ligne TVA manuelle','Indiquez le montant retenu et justifiez son traitement.',`${this.hidden(s)}<input type="hidden" name="key" value="${esc(l?.key||'')}">${fields}<div class="form-grid">${field('TVA retenue pour cette période','retained_tax',((l?.retained_tax||0)/100).toFixed(2),'number',true,true,`min="0" max="${source?l.tax/100:10000000}" step="0.01"`)}${field('Justification du traitement','reason',l?.reason||'','text',true,true)}</div>${l?`<button type="button" class="btn" data-action="vat-remove" data-key="${esc(l.key)}">Retirer cette ligne</button>`:''}`,'line','Enregistrer la revue',true);
  },
  settings(){const s=this.current();this.form('Crédit antérieur & notes',this.name(s),`${this.hidden(s)}<div class="form-grid">${field('Crédit antérieur à reprendre','opening_credit',(s.opening_credit/100).toFixed(2),'number',true,true,'min="0" max="10000000" step="0.01"')}${field('Origine et validation du crédit','credit_note',s.credit_note,'text',false,true)}${field('Note de préparation','note',s.note,'text',false,true)}</div><div class="notice">Le report est saisi et validé manuellement. Atlas ne reprend pas automatiquement le crédit d’une autre période.</div>`,'settings','Enregistrer les notes');},
  decision(action,key){
    const s=this.current(),approve=action==='approve',remove=action==='remove';
    this.form(approve?'Approuver cette préparation ?':remove?'Retirer cette ligne ?':s.status==='draft'?'Abandonner la préparation ?':'Annuler cette approbation ?',this.name(s),`${this.hidden(s)}${remove?`<input type="hidden" name="key" value="${esc(key)}">`:''}${approve?`<div class="notice">À payer estimé : ${currency(s.totals.payable)} · Crédit restant estimé : ${currency(s.totals.carryforward)}. L’instantané sera verrouillé. Cette action n’effectue aucun dépôt auprès de la DGI.</div><label class="vat-confirm"><input type="checkbox" name="confirmed" required> J’ai vérifié le périmètre, les traitements TVA, les montants et le crédit antérieur.</label>`:`<div class="notice">${remove?'Une source de la période retirée devra être réimportée avant approbation. Pour l’exclure du calcul, retenez plutôt zéro avec une justification.':'L’historique sera conservé. Une nouvelle préparation pourra être créée pour cette période. Vérifiez séparément les éventuels reports vers les périodes suivantes.'}</div>${field('Motif','reason','','text',true,true)}`}`,action,approve?'Confirmer l’approbation':remove?'Confirmer le retrait':'Confirmer l’annulation');
  },
  async mutate(action,data,form){
    if(this.saving)return;this.saving=true;AtlasPlatform.busy++;
    const company=AtlasPlatform.companyId,button=form?.querySelector('[type=submit]'),errorBox=form?.querySelector('.form-error');
    if(button)button.disabled=true;if(errorBox)errorBox.textContent='';
    try{
      const result=await api(`vat-${action}`,{...data,request_key:form?this.key:crypto.randomUUID()});
      if(company!==AtlasPlatform.companyId)return;
      this.selected=result.id;await refresh();if(company!==AtlasPlatform.companyId)return;
      if(form){dialog.close();main.focus({preventScroll:true});}
      toast(action==='import'?`${result.added} sources ajoutées · ${result.changed} actualisées.`:action==='approve'?'Préparation approuvée en interne.':'Enregistrement effectué.');
    }catch(error){if(errorBox&&form.isConnected)errorBox.textContent=error.message;else toast(error.message,true);}
    finally{this.saving=false;AtlasPlatform.busy--;if(button)button.disabled=false;}
  }
};
document.addEventListener('click',event=>{
  const b=event.target.closest('[data-action^="vat-"]');if(!b)return;
  event.preventDefault();event.stopImmediatePropagation();const v=AtlasVat;if(v.saving)return;
  const action=b.dataset.action.slice(4),s=v.current();
  ({new:()=>v.create(),open:()=>{v.selected=Number(b.dataset.id);render();},back:()=>{v.selected=null;render();},
    import:()=>v.mutate('import',{id:s.id,version:s.version}),manual:()=>v.line(),line:()=>v.line(b.dataset.key),
    settings:()=>v.settings(),approve:()=>v.decision('approve'),void:()=>v.decision('void'),remove:()=>v.decision('remove',b.dataset.key)})[action]?.();
});
document.addEventListener('submit',event=>{
  const form=event.target,name=form.getAttribute('id');if(!name?.startsWith('vat-'))return;
  event.preventDefault();event.stopImmediatePropagation();
  const data=Object.fromEntries(new FormData(form));if(name==='vat-approve-form')data.confirmed=data.confirmed==='on';
  AtlasVat.mutate(name.slice(4,-5),data,form);
});
